File: src/blindspot/blindspot/ros_bridge.py
"""Turning detector output into something the guard can score.

No ROS imports here on purpose: this is the part that can be wrong, so it is
testable without a ROS installation. The node in blindspot_ros/ is a thin
wrapper around it.

Two things this exists to get right.

ORDERING. The guard signal is a shoelace polygon area, which depends on the
ORDER of the points. A detector reports whatever order it happens to find, and
an order that crosses itself gives a small area - firing the guard for a
reason that has nothing to do with the geometry. Points are therefore ordered
by detection id, which reproduces the order the calibration used. Without ids
we fall back to an angular sort about the centroid, which is correct for a
convex arrangement and is reported so it can be seen in the diagnostic.

UNITS. The guard works in normalised image coordinates. Use the principal
point from CameraInfo when the camera is actually calibrated; only fall back
to (W-1)/2 when it is not, and say which was used. Note (W-1)/2, not W/2:
OpenGL maps NDC onto a continuous pixel range so the axis lands at W/2, while
OpenCV indexes pixel centres half a pixel lower.
"""

import numpy as np

from .units import pixels_to_normalised


def detection_center(det):
    """(x, y) pixel centre of a Detection2D, across vision_msgs versions.

    vision_msgs 4.x uses BoundingBox2D.center = vision_msgs/Pose2D, which has
    a `.position` with x and y. Earlier versions used geometry_msgs/Pose2D,
    with x and y directly on the centre. Duck-typed so both work and so this
    is testable without a ROS installation.
    """
    c = det.bbox.center
    pos = getattr(c, "position", None)
    if pos is not None:
        return float(pos.x), float(pos.y)
    return float(c.x), float(c.y)


def detection_label(det):
    """Stable id for a Detection2D, or None if the detector reports none.

    Prefers Detection2D.id, falls back to the first hypothesis class_id
    (vision_msgs 4.x) or result id (earlier).
    """
    did = getattr(det, "id", "")
    if did:
        return did
    for r in getattr(det, "results", []) or []:
        hyp = getattr(r, "hypothesis", None)
        cid = (getattr(hyp, "class_id", None) if hyp is not None
               else getattr(r, "id", None))
        if cid not in (None, ""):
            return cid
    return None


def angular_order(uv):
    """Indices that sort points counter-clockwise about their centroid."""
    uv = np.asarray(uv, dtype=float)
    c = uv.mean(axis=0)
    return np.argsort(np.arctan2(uv[:, 1] - c[1], uv[:, 0] - c[0]))


def order_features(uv, ids=None):
    """Order detections so the shoelace area means what it should.

    Returns (ordered_uv, ordered_ids, how) where `how` is "id" or "angular".
    """
    uv = np.asarray(uv, dtype=float)
    if ids is not None and len(ids) == len(uv):
        keys = []
        for i in ids:
            try:
                keys.append((0, float(i)))
            except (TypeError, ValueError):
                keys.append((1, str(i)))
        try:
            idx = np.array(sorted(range(len(uv)), key=lambda k: keys[k]))
        except TypeError:                     # mixed unorderable keys
            idx = np.array(sorted(range(len(uv)), key=lambda k: str(ids[k])))
        return uv[idx], [ids[i] for i in idx], "id"
    idx = angular_order(uv)
    return uv[idx], None, "angular"


def intrinsics_from_camera_info(K, width, height):
    """(fx, fy, cx, cy, source) from a CameraInfo K matrix.

    K is the row-major 9-vector [fx 0 cx; 0 fy cy; 0 0 1]. An uncalibrated
    camera publishes zeros, in which case there is nothing to use and the
    caller must supply a field of view instead.
    """
    K = np.asarray(K, dtype=float).reshape(3, 3)
    fx, fy, cx, cy = K[0, 0], K[1, 1], K[0, 2], K[1, 2]
    if fx > 0 and fy > 0:
        return float(fx), float(fy), float(cx), float(cy), "camera_info"
    return None, None, (width - 1) / 2.0, (height - 1) / 2.0, "uncalibrated"


def features_to_normalised(uv, fx, fy, cx, cy, ids=None):
    """Detector pixels -> ordered normalised coordinates for the guard."""
    ordered, ordered_ids, how = order_features(uv, ids)
    return pixels_to_normalised(ordered, fx, fy, cx, cy), ordered_ids, how


def ordering_is_suspect(uv, tol=0.98):
    """True if the given order encloses much less area than an angular sort.

    A cheap self-check: for points in a sane order the two agree. If they do
    not, the order crosses itself and the area - and therefore the guard
    decision - is not measuring the geometry.

    The reference is the polygon of the angular sort about the centroid, which
    is the convex hull only when the points are in convex position; for a
    concave arrangement it under-reports, so this errs towards staying quiet.
    """
    uv = np.asarray(uv, dtype=float)
    if uv.shape[0] < 3:
        return False

    def shoelace(p):
        x, y = p[:, 0], p[:, 1]
        return 0.5 * abs(np.dot(x, np.roll(y, -1)) - np.dot(y, np.roll(x, -1)))

    hull = uv[angular_order(uv)]
    a_hull = shoelace(hull)
    if a_hull <= 0:
        return False
    return shoelace(uv) < tol * a_hull

File: src/blindspot_cpp/src/ros_bridge.cpp
#include "blindspot/ros_bridge.hpp"

#include <algorithm>
#include <cmath>
#include <numeric>

namespace blindspot
{

std::string to_string(Ordering how)
{
  return how == Ordering::kById ? "id" : "angular";
}

std::string to_string(IntrinsicsSource src)
{
  switch (src) {
    case IntrinsicsSource::kCameraInfo: return "camera_info";
    case IntrinsicsSource::kFovParameter: return "fovy_deg parameter";
    default: return "uncalibrated";
  }
}

std::vector<int> angular_order(const Eigen::Ref<const Eigen::MatrixX2d> & uv)
{
  const Eigen::Index n = uv.rows();
  const Eigen::RowVector2d c = uv.colwise().mean();
  std::vector<double> angle(static_cast<size_t>(n));
  for (Eigen::Index i = 0; i < n; ++i) {
    angle[static_cast<size_t>(i)] = std::atan2(uv(i, 1) - c(1), uv(i, 0) - c(0));
  }
  std::vector<int> idx(static_cast<size_t>(n));
  std::iota(idx.begin(), idx.end(), 0);
  std::stable_sort(
    idx.begin(), idx.end(),
    [&angle](int a, int b) {return angle[static_cast<size_t>(a)] <
      angle[static_cast<size_t>(b)];});
  return idx;
}

namespace
{

/// Parses an id as a number. Ids that parse sort numerically and ahead of
/// any that do not.
bool numeric_id(const std::string & s, double * out)
{
  try {
    size_t used = 0;
    const double v = std::stod(s, &used);
    if (used != s.size()) {
      return false;
    }
    *out = v;
    return true;
  } catch (const std::exception &) {
    return false;
  }
}

Eigen::MatrixX2d take_rows(
  const Eigen::Ref<const Eigen::MatrixX2d> & uv, const std::vector<int> & idx)
{
  Eigen::MatrixX2d out(static_cast<Eigen::Index>(idx.size()), 2);
  for (size_t i = 0; i < idx.size(); ++i) {
    out.row(static_cast<Eigen::Index>(i)) = uv.row(idx[i]);
  }
  return out;
}

double shoelace(const Eigen::Ref<const Eigen::MatrixX2d> & p)
{
  const Eigen::Index n = p.rows();
  double cross = 0.0;
  for (Eigen::Index i = 0; i < n; ++i) {
    const Eigen::Index j = (i + 1) % n;
    cross += p(i, 0) * p(j, 1) - p(i, 1) * p(j, 0);
  }
  return 0.5 * std::abs(cross);
}

}  // namespace

OrderedFeatures order_features(
  const Eigen::Ref<const Eigen::MatrixX2d> & uv, const std::vector<std::string> & ids)
{
  const size_t n = static_cast<size_t>(uv.rows());
  if (!ids.empty() && ids.size() == n) {
    std::vector<int> idx(n);
    std::iota(idx.begin(), idx.end(), 0);
    std::stable_sort(
      idx.begin(), idx.end(),
      [&ids](int a, int b) {
        double va = 0.0;
        double vb = 0.0;
        const bool na = numeric_id(ids[static_cast<size_t>(a)], &va);
        const bool nb = numeric_id(ids[static_cast<size_t>(b)], &vb);
        if (na != nb) {
          return na;                        // numeric ids sort first
        }
        if (na) {
          return va < vb;
        }
        return ids[static_cast<size_t>(a)] < ids[static_cast<size_t>(b)];
      });
    std::vector<std::string> ordered_ids;
    ordered_ids.reserve(n);
    for (const int i : idx) {
      ordered_ids.push_back(ids[static_cast<size_t>(i)]);
    }
    return OrderedFeatures{take_rows(uv, idx), std::move(ordered_ids), Ordering::kById};
  }
  return OrderedFeatures{take_rows(uv, angular_order(uv)), {}, Ordering::kAngular};
}

CameraInfoIntrinsics intrinsics_from_camera_info(
  const std::array<double, 9> & k, int width, int height)
{
  const double fx = k[0];
  const double fy = k[4];
  const double cx = k[2];
  const double cy = k[5];
  if (fx > 0.0 && fy > 0.0) {
    return CameraInfoIntrinsics{
      Intrinsics{fx, fy, cx, cy}, IntrinsicsSource::kCameraInfo, true};
  }
  // Nothing usable was published. Only the fallback principal point is
  // meaningful; the caller must supply a field of view.
  return CameraInfoIntrinsics{
    Intrinsics{0.0, 0.0, (width - 1) / 2.0, (height - 1) / 2.0},
    IntrinsicsSource::kUncalibrated, false};
}

NormalisedFeatures features_to_normalised(
  const Eigen::Ref<const Eigen::MatrixX2d> & uv,
  const Intrinsics & k,
  const std::vector<std::string> & ids)
{
  OrderedFeatures ordered = order_features(uv, ids);
  return NormalisedFeatures{
    pixels_to_normalised(ordered.uv, k), std::move(ordered.ids), ordered.how};
}

bool ordering_is_suspect(const Eigen::Ref<const Eigen::MatrixX2d> & uv, double tol)
{
  if (uv.rows() < 3) {
    return false;
  }
  // Not the convex hull: the angular sort only coincides with it when the
  // points are in convex position.
  const Eigen::MatrixX2d sorted = take_rows(uv, angular_order(uv));
  const double reference = shoelace(sorted);
  if (reference <= 0.0) {
    return false;
  }
  return shoelace(uv) < tol * reference;
}

}  // namespace blindspot
